dali/operators/audio/mfcc/mfcc.cc

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/audio/mfcc/mfcc.h"
+#include "dali/core/static_switch.h"
+#include "dali/kernels/signal/dct/dct_cpu.h"
+#include "dali/kernels/common/utils.h"
+#include "dali/kernels/common/for_axis.h"
+#include "dali/pipeline/data/views.h"
+
+
+#define MFCC_SUPPORTED_TYPES (float)
+#define MFCC_SUPPORTED_NDIMS (2, 3, 4)
+
+static constexpr int kNumInputs = 1;
+static constexpr int kNumOutputs = 1;
+
+namespace dali {
+
+namespace detail {
+
+template <typename T, int Dims>
+void ApplyLifter(const kernels::OutTensorCPU<T, Dims> &inout, int axis, const T* lifter_coeffs) {
+  auto* data = inout.data;
+  auto shape = inout.shape;
+  auto strides = kernels::GetStrides(shape);
+  kernels::ForAxis(
+    data, data, shape.data(), strides.data(), shape.data(), strides.data(), axis, Dims,
+    [lifter_coeffs](
+      T *out_data, const T *in_data, int64_t out_size, int64_t out_stride,
+      int64_t in_size, int64_t in_stride) {
+        int64_t idx = 0;
+        assert(out_size == in_size);
+        assert(out_stride == in_stride);
+        for (int64_t k = 0; k < out_size; k++, idx += out_stride) {
+          out_data[idx] = lifter_coeffs[k] * in_data[idx];
+        }
+      });
+}
+
+}  // namespace detail
+
+DALI_SCHEMA(MFCC)
+    .DocStr(R"code(Mel Frequency Cepstral Coefficiencs (MFCC).
+Computes MFCCs from a mel spectrogram)code")
+    .NumInput(kNumInputs)
+    .NumOutput(kNumOutputs)
+    .AddOptionalArg("n_mfcc",
+      R"code(Number of MFCC coefficients)code",
+      20)
+    .AddOptionalArg("dct_type",
+      R"code(Discrete Cosine Transform type. Supported types are: 1, 2, 3, 4.
+The formulas used to calculate the DCT are equivalent to those described in
+https://en.wikipedia.org/wiki/Discrete_cosine_transform)code",
+      2)
+    .AddOptionalArg("normalize",
+      R"code(If true, the DCT will use an ortho-normal basis.
+Note: Normalization is not supported for `dct_type=1`.)code",
+      false)
+    .AddOptionalArg("axis",
+      R"code(Axis over which the transform will be applied.
+If not provided, the outer-most dimension will be used.)code",
+      0)
+    .AddOptionalArg("lifter",
+      R"code(Cepstral filtering (also known as `liftering`) coefficient.
+If `lifter>0`, the MFCCs will be scaled according to the following formula::
```

Comment:
Trailing double colon starts a pre-formatted block.
```suggestion
If `lifter>0`, the MFCCs will be scaled according to the following formula:
```

Reply:
I know, that's what @klecki suggested I can use to display equations and such

Reply:
Yes, I would stick with it.